enhancements/security/manual-cloud-credentials.md

```diff
@@ -0,0 +1,170 @@
+
+  * Providing a mechanism to disable the cloud credential operator, this mechanism already exists.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints [optional]
+
+This proposal requires the CCO to begin reporting Upgradable=False in a number of scenarios, unless an admin informs us they have taken action to reconcile new permissions and it is safe to upgrade.
+
+CLI tooling will be implemented to help users who wish to disable the CCO to perform their initial installations, update credentials prior to upgrade for release images available for update, and reconcile credentials against current CredentialsRequests in the cluster on an on-going basis. Users must take this manual step to clear the Upgradable=False condition if credentials have changed in any available update.
+
+Cloud credential operator reconcile code will be refactored to be a little more friendly to use as a vendored library, for use in the CLI.
+
+Note that when we say CCO is disabled, it is still running, just not minting credentials, and does not have access to the powerful "root" credentials for the cloud provider account.
+
+### Risks and Mitigations
+
+
+## Design Details
+
+Core of this proposal will be CLI tooling which vendors code from the credentials operator and allows for single use execution of the reconcile loop. Provising CLI tooling allows users to bypass the security concerns of storing powerful cloud credentials in the cluster and instead keep them entirely locally.
+
+### Upgradable False Condition
+
+A field will be added to the cloud credential operator ConfigMap (possibly becoming a separate API type in another enhancement) storing a list of Updates (from ClusterVersion.AvailableUpdates) that have been reconciled, confirming that we have minted credentials and it is safe to upgrade to this release.
+
+```yaml
+  reconciledAvailableUpdates:
+  - version: 4.4.1
+    image: quay.io/openshift-release-dev/ocp-release:4.4.1-x86_64
+  - version: 4.4.2
+    image: quay.io/openshift-release-dev/ocp-release:4.4.2-x86_64
+```
+
+Entries will be added to this list by the CLI tooling below, however an administrator could do this by hand if desired.
+
+If the CCO is enabled per normal best practice install, this field will be ignored, and should be absent.
+
+However if CCO is disabled, it will compare the ClusterVersion.Spec.AvailableUpdates to this list of ReconciledAvailableUpdates. If *any* available update is not present in the reconciled list, the CCO will report Upgradable=False.
+
+### Initial Installation
+
+```bash
+$ oc adm credentials reconcile --manifests=/path/to/installer/manifests
+```
+
+This command would parse the CredentialsRequests in the manifests directory, mint credentials in the cloud using the local users environment variables or credentials file, store the status on the CredentialsRequsts (which are yet to be created), write Secret manifests containing the cloud credentials, and disable the CCO.
+
+### Upgrade Reconciliation
```

Comment:
Personally i think we should,

1. degrade with clear action item when cco is disabled and the Credential Requests have changed.
we can achieve this by making the CredentialsRequests objects one _level_ before the cluster operator of credential operator, and storing hashes on actual requests to see if API permissions have changed.

2. allow users to list the changes given a release image or version (use the ClusterVersionStatus.AvailableUpdates) for that.

The current upgradeable False is just not a great UX that we can support right now, something we might need to support as a CVO enhancement for pre-flight checks.

Reply:
Upgradable=false until you tell us you've acted, if you've opted into disabling CCO seemed like a reasonable approach in the arch call back in March. It's not great UX but you've opted into this, and from what I've heard from PM these kinds of customers are ok with doing a little legwork. It clearly prevents us from letting someone enter an upgrade that is going to get stuck.

If we change approach and instead let all upgrades proceed, and mark CCO disabled if we detect a hash change on CredRequest.Spec, we end up with:

1. Doesn't CCO going degraded impact the upgrade itself? My understanding after the arch call was that this is the key thing we're trying to avoid, preventing users from starting an upgrade that won't complete.
2. It seems we're still open to the risk an SLO needed new perms and will go degraded during upgrade, getting the upgrade stuck (hasn't happened yet, 4.5 requires new machine-api perms, but the team thinks it won't degrade anything unless you use new features) Regardless the risk is still there.
3. We're always going to go degraded on any permissions change, regardless if it actually impacted anything or not
4. There's no way for a user to pre-bake things and get a clean upgrade.
5. How does a user clear this state? Same CLI tooling as proposed?

We were unaware of this preflight feature, I assume that is upgrade preflight checks? And is it per operator? Or something global in the CVO? In any case it sounds very interesting, so coming up with something to get us by until that lands would be worth discussing. We do need to deliver disabled CCO as a supported mechanism in 4.6 though, so I believe we need to do something in the meantime.